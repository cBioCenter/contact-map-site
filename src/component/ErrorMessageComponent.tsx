// ~contact-map-site~
// Error Message Component
// Renders an error message with details, if available.
// ~contact-map-site~

import { BioblocksPDB, CouplingContainer, IResidueMismatchResult } from 'bioblocks-viz';
import * as React from 'react';
import { Accordion, Message } from 'semantic-ui-react';

export interface IErrorMessageComponentProps {
  couplingScores: CouplingContainer;
  errorMsg: string;
  mismatches: IResidueMismatchResult[];
  pdbData?: BioblocksPDB;
}

export class ErrorMessageComponent extends React.Component<IErrorMessageComponentProps> {
  public static defaultProps = {
    couplingScores: new CouplingContainer(),
    errorMsg: '',
    mismatches: [],
  };

  constructor(props: IErrorMessageComponentProps) {
    super(props);
  }

  public render() {
    const { couplingScores, errorMsg, mismatches, pdbData } = this.props;

    return mismatches.length >= 1 || errorMsg.length >= 1 ? (
      <Message warning={true}>
        {mismatches.length >= 1 && pdbData ? (
          <div>
            <Message.Header>
              {`Residue numbering mismatch detected. Please upload a file to correct the position numbering differences.`}
              {<br />}
              {`EVCouplings and EVFold outputs this file in the `}
              <strong>OUTPUT</strong>
              {` directory.
                This file might be named similar to
                '${pdbData.name}.csv' or '${pdbData.name}.indextableplus'`}
            </Message.Header>
            <Message.List>{errorMsg}</Message.List>
            <Message.Content>
              <Accordion
                fluid={true}
                exclusive={false}
                defaultActiveIndex={[]}
                panels={[
                  this.renderSequenceAccordionMessage('PDB sequence', pdbData.sequence, mismatches, 'pdb'),
                  this.renderSequenceAccordionMessage(
                    'Couplings Score sequence',
                    couplingScores.sequence,
                    mismatches,
                    'coupling',
                  ),
                ]}
              />
            </Message.Content>
          </div>
        ) : (
          errorMsg
        )}
      </Message>
    ) : null;
  }

  protected renderSequenceAccordionMessage = (
    title: string,
    sequence: string,
    mismatches: IResidueMismatchResult[],
    pdbOrCoupling: 'pdb' | 'coupling',
  ) => {
    let startIndex = 0;
    let prevResno = mismatches[0].resno;
    const result = new Array<JSX.Element>();

    for (const mismatch of mismatches) {
      const aminoAcid = pdbOrCoupling === 'pdb' ? mismatch.pdbAminoAcid : mismatch.couplingAminoAcid;
      const resIndex = sequence.indexOf(aminoAcid.singleLetterCode, startIndex + (mismatch.resno - prevResno));

      if (resIndex >= 0) {
        result.push(
          <React.Fragment key={`mismatch-${mismatch.resno}`}>
            <span style={{ color: 'black', fontSize: '12px' }}>{sequence.substring(startIndex, resIndex)}</span>
            <span style={{ color: 'red', fontSize: '16px', textDecoration: 'underline' }}>
              {sequence.charAt(resIndex)}
            </span>
          </React.Fragment>,
        );

        prevResno = mismatch.resno + 1;
        startIndex = resIndex + 1;
      }
    }

    result.push(
      <span key={'mismatch-final'} style={{ color: 'black', fontSize: '12px' }}>
        {sequence.substring(startIndex)}
      </span>,
    );

    return {
      content: {
        content: <p style={{ width: '80%', wordBreak: 'break-word' }}>{result}</p>,
      },
      key: `panel-${title}`,
      title: {
        content: `${title} (${sequence.length} Amino Acids)`,
      },
    };
  };
}
