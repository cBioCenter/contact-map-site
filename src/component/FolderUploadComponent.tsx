import * as React from 'react';
// tslint:disable-next-line: import-name
import Dropzone, { DropEvent } from 'react-dropzone';
import { Icon, Label } from 'semantic-ui-react';

export interface IFolderUploadZoneComponent {
  style: React.CSSProperties;
  onDrop?(acceptedFiles: File[], rejectedFiles: File[], event: DropEvent): Promise<void>;
}

export class FolderUploadComponent extends React.Component<IFolderUploadZoneComponent, any> {
  public static defaultProps = {
    style: {
      border: '7px dashed lightblue',
      padding: '20px',
      textAlign: 'center' as const,
    },
  };

  constructor(props: IFolderUploadZoneComponent) {
    super(props);
  }

  public render() {
    const { onDrop, style } = this.props;

    return (
      <Dropzone onDrop={onDrop}>
        {({ getRootProps, getInputProps }) => {
          const { ref, ...rootProps } = getRootProps();

          return (
            <div
              ref={ref as React.RefObject<HTMLDivElement>}
              style={{ ...FolderUploadComponent.defaultProps.style, ...style }}
              {...rootProps}
            >
              <input {...getInputProps()} />
              <Label circular={true} size={'massive'}>
                <Icon name={'upload'} size={'large'} />
                Upload an EVCouplings folder or files!
              </Label>
            </div>
          );
        }}
      </Dropzone>
    );
  }
}
