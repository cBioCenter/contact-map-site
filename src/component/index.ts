export * from './ErrorMessageComponent';
export * from './FolderUploadComponent';
