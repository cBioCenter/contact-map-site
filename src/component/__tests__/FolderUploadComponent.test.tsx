import { shallow } from 'enzyme';
import * as React from 'react';

import { FolderUploadComponent } from '~contact-map-site~/component';

describe('FolderUploadComponent', () => {
  it('Should match existing snapshot when no props are provided.', () => {
    const wrapper = shallow(<FolderUploadComponent />);
    expect(wrapper).toMatchSnapshot();
  });
});
