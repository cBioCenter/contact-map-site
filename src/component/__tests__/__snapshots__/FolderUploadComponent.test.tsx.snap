// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderUploadComponent Should match existing snapshot when no props are provided. 1`] = `
<SidebarPushable>
  <Sidebar
    direction="left"
  >
    <Dropzone>
      <Component />
    </Dropzone>
  </Sidebar>
</SidebarPushable>
`;
